inputDataReactive <- reactive({
    bamfile <- my_values$bamfile

    possibleTag <- list(
        "integer" = c(
            "AM", "AS", "CM", "CP", "FI", "H0", "H1", "H2",
            "HI", "IH", "MQ", "NH", "NM", "OP", "PQ", "SM",
            "TC", "UQ"
        ),
        "character" = c(
            "BC", "BQ", "BZ", "CB", "CC", "CO", "CQ", "CR",
            "CS", "CT", "CY", "E2", "FS", "LB", "MC", "MD",
            "MI", "OA", "OC", "OQ", "OX", "PG", "PT", "PU",
            "Q2", "QT", "QX", "R2", "RG", "RX", "SA", "TS",
            "U2"
        )
    )
    library(Rsamtools)
    bamTop100 <- scanBam(BamFile(bamfile, yieldSize = 100),
        param = ScanBamParam(tag = unlist(possibleTag))
    )[[1]]$tag
    ntags <- names(bamTop100)[lengths(bamTop100) > 0]
    ntags

    outPath <- tempdir()
    my_values$outPath <- outPath


    ## shift the coordinates of 5'ends of alignments in the bam file
    library(TxDb.Hsapiens.UCSC.hg19.knownGene)
    ## if you don't have an available TxDb, please refer
    ## GenomicFeatures::makeTxDbFromGFF to create one from gff3 or gtf file.
    # seqlev <- "chr1" ## subsample data for quick run
    # seqinformation <- seqinfo(TxDb.Hsapiens.UCSC.hg19.knownGene)
    # which <- as(seqinformation[seqlev], "GRanges")
    # gal <- readBamFile(bamfile, tag = ntags, which = which, asMates = TRUE, bigFile = TRUE)
    # shiftedBamfile <- file.path(outPath, "shifted.bam")
    # gal1 <- shiftGAlignmentsList(gal, outbam = shiftedBamfile)
    # library(input$tx_db_input, character.only = T)
    # txs <- transcripts(get(input$tx_db_input))

    library(TxDb.Hsapiens.UCSC.hg19.knownGene)
    ## if you don't have an available TxDb, please refer
    ## GenomicFeatures::makeTxDbFromGFF to create one from gff3 or gtf file.
    seqlev <- "chr1" ## subsample data for quick run
    seqinformation <- seqinfo(TxDb.Hsapiens.UCSC.hg19.knownGene)
    which <- as(seqinformation[seqlev], "GRanges")
    gal <- readBamFile(bamfile, tag = ntags, which = which, asMates = TRUE, bigFile = TRUE)
    shiftedBamfile <- file.path(outPath, "shifted.bam")
    gal1 <- shiftGAlignmentsList(gal, outbam = shiftedBamfile)

    return(list("gal1" = gal1, "txs" = txs))
})

output$plot_pt_score <- renderPlot({
    # files will be output into outPath


    gal1 <- inputDataReactive()$gal1
    txs <- inputDataReactive()$txs

    pt <- PTscore(gal1, txs)
    plot(pt$log2meanCoverage, pt$PT_score,
        xlab = "log2 mean coverage",
        ylab = "Promoter vs Transcript"
    )
})


output$plot_nfr_score <- renderPlot({
    nfr <- NFRscore(gal1, txs)
    plot(nfr$log2meanCoverage, nfr$NFR_score,
        xlab = "log2 mean coverage",
        ylab = "Nucleosome Free Regions score",
        main = "NFRscore for 200bp flanking TSSs",
        xlim = c(-10, 0), ylim = c(-5, 5)
    )
})


output$plot_tssre_score <- renderPlot({
    gal1 <- inputDataReactive()$gal1
    txs <- inputDataReactive()$txs
    tsse <- TSSEscore(gal1, txs)
    tsse$TSSEscore
    plot(100 * (-9:10 - .5), tsse$values,
        type = "b",
        xlab = "distance to TSS",
        ylab = "aggregate TSS score"
    )
})


output$plotCumulativePercentage <- renderPlot({

    gal1 <- inputDataReactive()$gal1
    txs <- inputDataReactive()$txs
    bamfile <- my_values$bamfile
    outPath <- my_values$outPath


    library(BSgenome.Hsapiens.UCSC.hg19)
    library(phastCons100way.UCSC.hg19)
    ## run program for chromosome 1 only
    txs <- txs[seqnames(txs) %in% "chr1"]
    genome <- Hsapiens
    ## split the reads into NucleosomeFree, mononucleosome,
    ## dinucleosome and trinucleosome.
    ## and save the binned alignments into bam files.
    objs <- splitGAlignmentsByCut(gal1,
        txs = txs, genome = genome, outPath = outPath,
        conservation = phastCons100way.UCSC.hg19
    )
    ## list the files generated by splitGAlignmentsByCut.
    print(dir(outPath))

    # objs <- splitBam(bamfile, tag=ntags, outPath=outPath,
    #              txs=txs, genome=genome,
    #              conservation=phastCons100way.UCSC.hg19)

    library(ChIPpeakAnno)



    seqinformation <- seqinfo(TxDb.Hsapiens.UCSC.hg19.knownGene)
    outPath <- my_values$outPath
    bamfiles <- file.path(
        outPath,
        c(
            "NucleosomeFree.bam",
            "mononucleosome.bam",
            "dinucleosome.bam",
            "trinucleosome.bam"
        )
    )
    print(bamfiles)
    ## Plot the cumulative percentage of tag allocation in nucleosome-free
    ## and mononucleosome bam files.
    cumulativePercentage(bamfiles[1:2], as(seqinformation["chr1"], "GRanges"))
})


output$plot_featureAlignedHeatmap <- renderPlot({

    bamfiles <- file.path(outPath,
                     c("NucleosomeFree.bam",
                     "mononucleosome.bam",
                     "dinucleosome.bam",
                     "trinucleosome.bam"))


    gal1 <- inputDataReactive()$gal1
    txs <- inputDataReactive()$txs
    bamfile <- my_values$bamfile
    outPath <- my_values$outPath

    TSS <- promoters(txs, upstream=0, downstream=1)
    TSS <- unique(TSS)
    ## estimate the library size for normalization
    (librarySize <- estLibSize(bamfiles))
    NTILE <- 101
    dws <- ups <- 1010
    sigs <- enrichedFragments(
        gal = objs[c(
            "NucleosomeFree",
            "mononucleosome",
            "dinucleosome",
            "trinucleosome"
        )],
        TSS = TSS,
        librarySize = librarySize,
        seqlev = seqlev,
        TSS.filter = 0.5,
        n.tile = NTILE,
        upstream = ups,
        downstream = dws
    )
    ## log2 transformed signals
    sigs.log2 <- lapply(sigs, function(.ele) log2(.ele + 1))
    # plot heatmap

    output$plot_signals <- renderPlot({

        ## get signals normalized for nucleosome-free and nucleosome-bound regions.
        out <- featureAlignedDistribution(sigs, 
                                    reCenterPeaks(TSS, width=ups+dws),
                                    zeroAt=.5, n.tile=NTILE, type="l", 
                                    ylab="Averaged coverage")
        ## rescale the nucleosome-free and nucleosome signals to 0~1
        range01 <- function(x){(x-min(x))/(max(x)-min(x))}
        out <- apply(out, 2, range01)
        matplot(out, type="l", xaxt="n", 
                xlab="Position (bp)", 
                ylab="Fraction of signal")
        axis(1, at=seq(0, 100, by=10)+1, 
            labels=c("-1K", seq(-800, 800, by=200), "1K"), las=2)
        abline(v=seq(0, 100, by=10)+1, lty=2, col="gray")
    })

    featureAlignedHeatmap(sigs.log2, reCenterPeaks(TSS, width = ups + dws),
        zeroAt = .5, n.tile = NTILE
    )
})


output$plot_Footprints <- renderPlot({
    shiftedBamfile <- file.path(outPath, "shifted.bam")
    library(MotifDb)
    CTCF <- query(MotifDb, c("CTCF"))
    CTCF <- as.list(CTCF)
    print(CTCF[[1]], digits=2)

    sigs <- factorFootprints(shiftedBamfile, pfm=CTCF[[1]], 
                         genome=genome, ## Don't have a genome? ask ?factorFootprints for help
                         min.score="90%", seqlev=seqlev,
                         upstream=100, downstream=100)
})

output$plot_featureAlignedHeatmap <- renderPlot({
    shiftedBamfile <- file.path(outPath, "shifted.bam")
    library(MotifDb)
    CTCF <- query(MotifDb, c("CTCF"))
    CTCF <- as.list(CTCF)
    print(CTCF[[1]], digits=2)

    sigs <- factorFootprints(shiftedBamfile, pfm=CTCF[[1]], 
                         genome=genome, ## Don't have a genome? ask ?factorFootprints for help
                         min.score="90%", seqlev=seqlev,
                         upstream=100, downstream=100)

    output$plot_segmentation <-  renderPlot({
        sigs$Profile.segmentation
    })

    output$plot_vp <- renderPlot({
        vp <- vPlot(shiftedBamfile, pfm=CTCF[[1]], 
            genome=genome, min.score="90%", seqlev=seqlev,
            upstream=200, downstream=200, 
            ylim=c(30, 250), bandwidth=c(2, 1))

         output$plot_distanceDyad <- renderPlot({
         distanceDyad(vp, pch=20, cex=.5)
         })
    })


   
    

    featureAlignedHeatmap(sigs$signal, 
                      feature.gr=reCenterPeaks(sigs$bindingSites,
                                               width=200+width(sigs$bindingSites[1])), 
                      annoMcols="score",
                      sortBy="score",
                      n.tile=ncol(sigs$signal[[1]]))

})



